src/anyp/Uri.cc

```diff
@@ -411,23 +476,58 @@ AnyP::Uri::parse(const HttpRequestMethod& method, const SBuf &rawUrl)
                 return false;
             *dst = '\0';
 
-            // We are looking at path-abempty.
-            if (*src != '/') {
-                // path-empty, including the end of the `src` c-string cases
-                urlpath[0] = '/';
-                dst = &urlpath[1];
-            } else {
+            bool chopped = false;
+            if (*src == '/') {
                 dst = urlpath;
-            }
-            /* Then everything from / (inclusive) until \r\n or \0 - that's urlpath */
-            for (; i < l && *src != '\r' && *src != '\n' && *src != '\0'; ++i, ++src, ++dst) {
-                *dst = *src;
+                /* Then everything from / (inclusive) until '?', '#', '\r\n' or '\0' - that's urlpath */
+                for (; i < l && *src != '?' && *src != '#' && *src != '\r' && *src != '\n' && *src != '\0'; ++i, ++src, ++dst) {
+                    *dst = *src;
+                }
+                *dst = '\0';
+                chopped = StripAnyWsp(urlpath);
+                rfc1738_unescape(urlpath);
+            } else {
+                // We should be looking at path-abempty. relative-path is not supported (yet).
+                urlpath[0] = '/';
+                urlpath[1] = '\0';
             }
 
             /* We -could- be at the end of the buffer here */
             if (i > l)
                 return false;
-            *dst = '\0';
+
+            if (*src == '?') {
+                dst = foundQuery;
+                ++src; // skip '?' delimiter
+                ++i; // keep track of bytes 'consumed' from src
+                /* Then everything from '?' (exclusive) until '#', '\r\n' or '\0' - that's query */
+                for (; i < l && *src != '#' && *src != '\r' && *src != '\n' && *src != '\0'; ++i, ++src, ++dst) {
+                    *dst = *src;
+                }
+                *dst = '\0';
+                if (chopped)
+                    *foundQuery = '\0';
+                else
+                    chopped = StripAnyWsp(foundQuery);
+                rfc1738_unescape(foundQuery);
+            }
+            if (i > l)
+                return false;
+
+            if (*src == '#') {
+                ++src; // skip '#' delimiter
+                ++i; // keep track of bytes 'consumed' from src
+                /* Then everything from '#' (exclusive) until '\r\n' or '\0' - that's fragment */
+                for (; i < l && *src != '\r' && *src != '\n' && *src != '\0'; ++i, ++src) {
+                    ; // fragment component is not expected in network protocols. discard for now.
```

Comment:
If official parse() code does not discard fragments, then this PR cannot use a "for now" excuse to start discarding them; most likely, this PR should not discard fragments at all, but we can discuss better excuses if really needed.